ion-schema/src/isl/version_based_isl/mod.rs

```diff
@@ -0,0 +1,480 @@
+//!     "my_type_name".to_owned(),
+//!     vec![
+//!         // represents the `type: int` constraint
+//!         IslConstraint::type_constraint(
+//!             IslTypeArgument::named("int")
+//!         ),
+//!         // represents `all_of` with anonymous type `{ type: bool }` constraint
+//!         IslConstraint::all_of(
+//!             vec![
+//!                 IslTypeArgument::anonymous(
+//!                     vec![
+//!                         IslConstraint::type_constraint(
+//!                             IslTypeArgument::named("bool")
+//!                         )
+//!                     ]
+//!                 )
+//!             ]
+//!         ),
+//!         // represents some open content which will be ignored while doing validation
+//!         IslConstraint::open_content("open_content".to_owned(), Element::string("This is an open content!"))
+//!     ]
+//! );
+//!
+//! // create an ISL schema using above IslType
+//! let isl_schema_header = IslSchemaHeader::<IslV1_0>::new(vec![], vec![]);
+//! let mut  isl_schema = IslSchema::new("my_schema", vec![isl_type], vec![], vec![]);
+//! isl_schema = isl_schema.with_schema_header(isl_schema_header);
+//!
+//! assert_eq!(isl_schema.types().len(), 1);
+//! assert_eq!(isl_schema.types()[0].name(), &Some("my_type_name".to_owned()));
+//! assert_eq!(isl_schema.types()[0].constraints().len(), 3);
+//! // verify that the last constraint is actually an open content field
+//! assert!(matches!(isl_schema.types()[0].constraints()[2], IslConstraint::Unknown(_)));
+//! ```
+//!
+//! ## Example usage of `version_based_isl` module to create an `IslSchema` using `IslType` for ISL 2.0:
+//! ```
+//! use ion_rs::element::Element;
+//! use ion_schema::isl::version_based_isl::constraints::IslConstraint;
+//! use ion_schema::isl::version_based_isl::{IslSchema, IslSchemaHeader, IslTypeArgument, IslV2_0, UserReservedFields};
+//! use ion_schema::isl::version_based_isl::constraints::open_content::OpenContent;
+//! use crate::ion_schema::isl::version_based_isl::IslType;
+//! use crate::ion_schema::isl::version_based_isl::constraints::r#type::Type;
+//!
+//! // below code represents an ISL type:
+//! // type:: {
+//! //      name:my_type_name,
+//! //      type: int,
+//! //  }
+//! let isl_type: IslType<IslV2_0> = IslType::named(
+//!     // represents the `name` of the defined type
+//!     "my_type_name".to_owned(),
+//!     vec![
+//!         // represents the `type: int` constraint
+//!         IslConstraint::type_constraint(
+//!             IslTypeArgument::named("int")
+//!         ),
+//!     ]
+//! );
+//!
+//! let mut isl_schema_header =  IslSchemaHeader::<IslV2_0>::new(vec![], vec![]); // no open content or imports are there in schema header
+//! // Performing the following operation with ISL 1.0 schema would result in a compile time error.
+//! isl_schema_header = isl_schema_header.with_user_reserved_fields(vec!["foo".to_string(), "bar".to_string()], vec![], vec![]);
```

Comment:
I don't think `cargo fmt` will ever change the formatting of doc examples. Consider manually wrapping the text if it gets to be longer than 120 or so columns.